### ISSUE TITLE
Making a map of the mask returned by get_bpt in a multi-panel plot produces no plot

### ISSUE THREAD

**Issue by KateRowlands** [sdss/marvin#320]
The following code produces no plot, just writes out a matplotlib object :
```

from matplotlib import colors

#List of targets
select = np.where((plateifu == '8987-9101') | (plateifu == '8555-3701') | 
                          (plateifu == '8440-6104') | (plateifu == '8080-3702 ') | (plateifu == '8547-12705') )

n_select = np.round(len(plateifu[select])/2)

cmap_bpt = colors.ListedColormap(['cyan', 'green', 'red', 'magenta', 'grey'])
bounds_bpt=[0,1,2,3,4,5,6]
norm_bpt = colors.BoundaryNorm(bounds_bpt, cmap_bpt.N)

cols=4

fig, axs = plt.subplots(nrows=n_select, ncols=cols, figsize=(3*cols, 2.5*n_select))
fig.subplots_adjust(hspace=0.26, wspace=.22)

for i in range(n_select):
    for j in range(2):

        ind = select[0][(2*i+j)]
        
        maps = Maps(plateifu=plateifu[ind])
        ha = maps['emline_gflux_ha_6564']
        nocov = ha.pixmask.get_mask('NOCOV')

        mask = nocov
        
        #Plot SDSS image
        image = showImage(plateifu=plateifu[ind], show_image=False)
        img = axs[i, j*2].imshow(image)
        axs[i, j*2].set_title(str(plateifu[ind]), fontsize=13)
        axs[i, j*2].get_xaxis().set_visible(False) #Suppress tick labels
        axs[i, j*2].get_yaxis().set_visible(False)

        #BPT plot
        bpt_mask = maps.get_bpt(use_oi=False, show_plot=False, return_figure=False)
        
        # Gets the masks for NII/Halpha        
        sf = bpt_mask['sf']['global']
        comp = bpt_mask['comp']['global']
        agn = bpt_mask['agn']['global']
        seyfert = bpt_mask['seyfert']['global']
        liner = bpt_mask['liner']['global']
        ambiguous = bpt_mask['ambiguous']['global']
        invalid = bpt_mask['invalid']['global']
        
        bpt = np.empty((ha.shape[0], ha.shape[1]))
        bpt[sf] = 1
        bpt[comp] = 2
        bpt[seyfert] = 3
        bpt[liner] = 4
        bpt[ambiguous] = 5
        
        fig, ax, cb = mapplot.plot(value=bpt, mask=mask, \
                     cmap=cmap_bpt, fig=fig, ax=axs[i, (j*2)+1], cbrange=(0,5), return_cb=True)
        cb.remove()
                                
plt.savefig(fig_dir+'BPT_multipanel_test.pdf', bbox_inches='tight')
plt.show()
```

Produces:
<matplotlib.figure.Figure at 0x10abc2fd0>

If I replace the line bpt_mask = maps.get_bpt(use_oi=False, show_plot=False, return_figure=False) with an array which already has a BPT mask saved from file, then the plot is produced as expected.

**Comment by bretthandrews** [comment]
Edited version of example code complete with imports/data fetching:

```
import numpy as np
from matplotlib import colors
import matplotlib.pyplot as plt

from marvin.tools.maps import Maps
from marvin.utils.general.images import showImage
import marvin.utils.plot.map as mapplot
from marvin import config
config.forceDbOff()


# List of targets
plateifu = np.array(['8987-9101', '8555-3701', '8440-6104', '8080-3702', '8547-12705'])
select = np.where((plateifu == '8987-9101') | (plateifu == '8555-3701') |
                  (plateifu == '8440-6104') | (plateifu == '8080-3702') |
                  (plateifu == '8547-12705'))

n_select = int(np.round(len(plateifu[select]) / 2))

cmap_bpt = colors.ListedColormap(['cyan', 'green', 'red', 'magenta', 'grey'])
bounds_bpt = [0, 1, 2, 3, 4, 5, 6]
norm_bpt = colors.BoundaryNorm(bounds_bpt, cmap_bpt.N)

cols = 4

fig, axs = plt.subplots(nrows=n_select, ncols=cols, figsize=(cols, n_select))
fig.subplots_adjust(hspace=0.26, wspace=.22)

for i in range(n_select):
    for j in range(2):

        ind = select[0][(2 * i + j)]

        maps = Maps(plateifu=plateifu[ind])
        ha = maps['emline_gflux_ha_6564']
        nocov = ha.pixmask.get_mask('NOCOV')

        mask = nocov

        # Plot SDSS image
        image = showImage(plateifu=plateifu[ind], show_image=False)
        img = axs[i, j * 2].imshow(image)
        axs[i, j * 2].set_title(str(plateifu[ind]), fontsize=13)
        axs[i, j * 2].get_xaxis().set_visible(False)  # Suppress tick labels
        axs[i, j * 2].get_yaxis().set_visible(False)

        # BPT plot
        bpt_mask = maps.get_bpt(use_oi=False, show_plot=False, return_figure=False)

        # Gets the masks for NII/Halpha
        sf = bpt_mask['sf']['global']
        comp = bpt_mask['comp']['global']
        agn = bpt_mask['agn']['global']
        seyfert = bpt_mask['seyfert']['global']
        liner = bpt_mask['liner']['global']
        ambiguous = bpt_mask['ambiguous']['global']
        invalid = bpt_mask['invalid']['global']

        bpt = np.empty((ha.shape[0], ha.shape[1]))
        bpt[sf] = 1
        bpt[comp] = 2
        bpt[seyfert] = 3
        bpt[liner] = 4
        bpt[ambiguous] = 5

        fig, ax, cb = mapplot.plot(value=bpt, mask=mask, cmap=cmap_bpt, fig=fig,
                                   ax=axs[i, (j * 2) + 1], cbrange=(0, 5), return_cb=True)
        cb.remove()
```

**Comment by bretthandrews** [comment]
I think I've narrowed down this issue to  `turn_off_ion()` in  `marivn.utils.general.general.py`. The problem is that calling `plt.close(fignum)` and/or `plt.close()` (see block below) closes the figure object for the multi-panel plot. I believe that they might have been intended to close dummy bpt figures. The plotting code (above) works if those `plt.close()` calls are removed. @albireox, can you comment as to what circumstances actually require the `plt.close()` calls? When I called `maps.get_bpt(use_oi=False, show_plot=False, return_figure=False)` without the multi-panel figure existing, it behaved properly with those `plt.close()` calls commented out.

```
    if show_plot:
        plt.ioff()
        plt.show()
    else:
        fignum = plt.get_fignums()
        if fignum:
            plt.close(fignum[0])
        else:
            plt.close()
```

**Comment by havok2063** [comment]
I needed this when implementing the scatter and histogram plotting from the query results.  I can't remember the exact circumstances but I think I was getting empty figures when `show_plot` was set to False, that should have been closed.  This code made sure that `show_plot=False` actually does not create and render an empty figure container.

**Comment by bretthandrews** [comment]
This code closes a figure that was created, so maybe there's a way to prevent the opening of an unwanted figure in the first place.

**Comment by havok2063** [comment]
@bretthandrews I'm going to ping you on this to see if you can take a look at it

**Comment by bretthandrews** [comment]
After commenting out the `plt.close(fignum)` and `plt.close()` lines (see this [comment](https://github.com/sdss/marvin/issues/320#issuecomment-364482526)), I can make a query results histogram and plot without showing the plots (and no error).

```
from marvin import config
config.mode = 'remote'
config.forceDbOff()
from marvin.tools.query import Query
myfilter = 'nsa.z < 0.1'
q = Query(searchfilter=myfilter, returnparams=['cube.ra', 'cube.dec'])
r = q.run()
h = r.hist('nsa.z', show_plot=False)
p = r.plot('cube.ra', 'cube.dec', show_plot=False)
```

@havok2063 can you reproduce the reason why we needed those `plt.close()`'s?

**Comment by havok2063** [comment]
Ok, in reproducing it I also remembered the problem.  The problem is without those lines, the next time you plot with `show_plot=True`, it opens up all previous plots.  

This is with MPL-4 on my local db but any release will do. 
```
from marvin import config
from marvin.tools.query import Query
config.setRelease('MPL-4')
config.switchSasUrl('local')
q = Query(searchfilter='nsa.z < 0.1', returnparams=['absmag_g_r'])
r = q.run()
# creates one plot
out = r.plot('nsa.z', 'absmag_g_r')
# creates no plot
out = r.plot('nsa.z', 'absmag_g_r', show_plot=False)
# creates two plots instead of one
out = r.plot('nsa.z', 'absmag_g_r')
```
I needed those lines to disable that.

**Comment by bretthandrews** [comment]
Fixed in 75851bd47f77240f6e3aafbaed5eed62b8f1ca9b. The `turn_off_ion` context manager now checks which figures existed beforehand and then cleans up any new plots that were spawned if `show_plot=False`.
